### PR TITLE
A bit stronger PHI simplification

### PR DESCRIPTION
**Pull request by Pheeck**
Don't know if this helps. I saw an opportunity to make simplify_phi() use a little bit stronger but just as fast optimization. I'm currently studying the project. In the future I may make more significant contributions.

### CODE REVIEW

File: lib/simplify_insts.cpp [hunk 1]
```diff
@@ -672,13 +672,21 @@ Instruction *simplify_memory(Instruction *inst)
 
 Instruction *simplify_phi(Instruction *phi)
 {
-  // A phi node can be replaced with an argument if all of them are identical.
-  Instruction *inst = phi->phi_args[0].inst;
+  // If phi only references itself or one other value, it can be replaced by
+  // that value, e.g. _1 = PHI<_1, _2, _1>;
```
Comment (kristerw) [inline lib/simplify_insts.cpp]: Please use the IR syntax used by parse_ir(), Module::print(), etc.

File: lib/simplify_insts.cpp [hunk 2]
```diff
@@ -672,13 +672,21 @@ Instruction *simplify_memory(Instruction *inst)
...
+  Instruction *inst = NULL;
   for (auto phi_arg : phi->phi_args)
     {
-      if (phi_arg.inst != inst)
-	return phi;
+      if (phi_arg.inst != phi)
+	{
+	  if (inst == NULL)
+	    inst = phi_arg.inst;
+	  else
```
Comment (kristerw) [inline lib/simplify_insts.cpp]: This does not look right to me... Doesn't this fail to optimize e.g.:
```
%2 = phi [ %1, .1], [ %1, .2], [ %1, .3]
```
as `inst` is not `NULL` when we come to the 2nd argument?

File: lib/simplify_insts.cpp [hunk 3]
```diff
@@ -672,13 +672,21 @@ Instruction *simplify_memory(Instruction *inst)
...
+  // that value, e.g. _1 = PHI<_1, _2, _1>;
+
+  Instruction *inst = NULL;
   for (auto phi_arg : phi->phi_args)
     {
-      if (phi_arg.inst != inst)
-	return phi;
+      if (phi_arg.inst != phi)
+	{
+	  if (inst == NULL)
```
Comment (kristerw) [inline lib/simplify_insts.cpp]: `NULL` -> `nullptr` (Or just use `if (inst)` etc.)

**Comment by Pheeck** [comment]
@kristerw Thanks for the review. I made adjustments according to your comments. You're right that the implementation wasn't correct. I think it is fine now.

**Comment by kristerw** [comment]
Looks good. Thanks!

**Pull request merged**